Simplify ref/ assembly generation

dougbu opened the pull request:
- followup 1/2 for 5266918ed2be
- correct the Razor.Tools project
  - `%(Reference.Version)` metadata does not bleed through into `@(PackageReference)` items
  - much more work to do so than to add this special case
- remove RTMVersions project and use RepoTasks instead
  - make it an error if RepoTasks is not restored before anything else builds
- add items and properties for System.Security.AccessControl

nits:
- remove invalid (ignored) metadata in Directory.Build.props and AzureAppServices.SiteExtension project
- improve / extend a couple of comments

dougbu committed:
Simplify ref/ assembly generation
- followup 1/2 for 5266918ed2be
- correct the Razor.Tools project
  - `%(Reference.Version)` metadata does not bleed through into `@(PackageReference)` items
  - much more work to do so than to add this special case
- remove RTMVersions project and use RepoTasks instead
  - make it an error if RepoTasks is not restored before anything else builds
- add items and properties for System.Security.AccessControl

nits:
- remove invalid (ignored) metadata in Directory.Build.props and AzureAppServices.SiteExtension project
- improve / extend a couple of comments

dougbu commented on src/Razor/Microsoft.AspNetCore.Razor.Tools/src/Microsoft.AspNetCore.Razor.Tools.csproj:
```diff
@@ -37,7 +37,7 @@
 
   <ItemGroup>
-    <Reference Include="Newtonsoft.Json" Version="$(Razor_NewtonsoftJsonPackageVersion)" />
+    <PackageReference Include="Newtonsoft.Json" Version="$(Razor_NewtonsoftJsonPackageVersion)" AllowExplicitReference="true" />
```
@NTaylorMullen @pranavkm @Pilchie the project currently builds against NewtonSoft.Json 12.0.2, not 9.0.1 in the 'master' and 'release/5.0-preview8' branches. Will that break customers who upgrade to the P8 SDK❔ If yes, I'll copy this one-liner into a PR for the p8 branch.

pranavkm replied:
This is a tool and is not referenced by user projects.

dougbu replied:
Then why are there really strong warnings in https://github.com/dotnet/aspnetcore/blob/6cef541d5fa9162af8a25a8d95639b392d2fd259/eng/Versions.props#L265-L267 ❔  If that's irrelevant, I could just remove the `$(Razor_NewtonsoftJsonPackageVersion)` property and this (one and only) attempted use.

dougbu replied:
@NTaylorMullen @JunTaoLuo what's the history here i.e. why was it considered necessary❔ @JunTaoLuo brought the comments over from aspnetcore-tooling but don't see anything beyond https://github.com/dotnet/aspnetcore/pull/21581/files#r428789251 about the block I quoted.

JunTaoLuo replied:
I'm not sure of the historical reason why this was done. i just duplicated the logic and comment from tooling. Maybe @pranavkm or @ajaybhargavb can explain further.

dougbu replied:
/ping @pranavkm and @NTaylorMullen is `$(Razor_NewtonsoftJsonPackageVersion)` still relevant❔ I'd rather remove the exception than make the current change if it'll work.

NTaylorMullen replied:
Those warnings are from the tooling time where our other tooling based binaries require a host bound newtonsoft.json version. This tool as pranav said is not referenced by user projects and can have whatever version we feel fit.

pranavkm reviewed (approved):
Razor TOols changes look fine

dougbu commented on eng/Versions.props:
```diff
@@ -163,6 +165,14 @@
+
+    If testing this configuration prior to servicing, update the versions of dependencies too. E.g. change
+    `$(SystemSecurityPrincipalWindowsV0PackageVersion)` if you change `$(SystemSecurityAccessControlV0PackageVersion)`
+    because System.Security.AccessControl will otherwise be loadable. This should not be necessary in servicing.
```
@ericstj when I updated only `$(SystemSecurityAccessControlV0PackageVersion)` in testing, the _newer_ (non-RTM) copy of System.Security.AccessControl.dll ended up in our targeting pack. I believe the newer assembly was also used in most compilations. It was difficult to tell exactly why but my guess is the newer System.Security.Principal.Windows.dll was somehow considered incompatible and conflict resolution chose the newer version of both. (The System.Security.Principal.Windows package is a direct dependency of a few of our packages.)

Will we hit something similar if dotnet/runtime ships System.Security.AccessControl 5.0.1 but leaves  System.Security.Principal.Windows at 5.0.0❔ Put another way, will dotnet/runtime build but not ship a newer  System.Security.Principal.Windows.dll and use that when compiling System.Security.AccessControl.dll❔ If yes, we'll need to be _really_ careful here (perhaps adding further workarounds) and I shouldn't include the "not be necessary" words.

ericstj replied:
There’s nothing fancy here. It just chooses the newer version. You’ll only see things go app local if you use a newer package than what’s in the shared framework you target.

dougbu replied:
We don't generally use our own part of the shared Fx when building and the example assemblies are in Microsoft.AspNetcore.App.Ref, not Microsoft.NETCore.App.Ref. So, does "nothing fancy" help us with the corner case I describe❔

I'd like to know we can rev one without the other reliably before we get to 5.0.1. (And, these packages / assemblies have a few other inter-dependencies.)

dougbu replied:
Unless someone objects, I'm going to assume the worst case here would be silently compiling against 5.0.x (newer) versions of these packages when servicing. Will file an issue about testing that `%(RTMVersion)` metadata is respected. That test will likely be a bit painful since we won't be building our targeting pack when servicing and checks will require looking at all references in our implementation assemblies for the relevant assemblies then checking their `[AssemblyInformationalVersion()]` attributes against the expected package version.

dougbu commented:
Side question: I see `ExceptionHandlerTest.ClearsResponseBuffer_BeforeRequestIsReexecuted(...)` failing in https://dev.azure.com/dnceng/public/_build/results?buildId=737724 (a build of master this morning) and https://dev.azure.com/dnceng/public/_build/results?buildId=738552 (the validation build for this PR). @JunTaoLuo is this a newly-flaky test you're about to quarantine❔

dougbu commented:
/azp run

wtgodbe commented on eng/targets/ResolveReferences.targets:
```diff
@@ -205,21 +205,23 @@
 
   <!--
-    Muck with @(ResolvedCompileFileDefinitions) items between generation and use in order to compile against RTM lib/
+    Change @(ResolvedCompileFileDefinitions) items between generation and use in order to compile against RTM lib/
```
Good change 😆

wtgodbe commented on Directory.Build.props:
```diff
@@ -151,9 +151,7 @@
-      PrivateAssets="All"
-      Version="$(InternalAspNetCoreAnalyzersPackageVersion)"
-      IsImplicitlyDefined="true" />
+      PrivateAssets="All" />
```
We get the versions from `dependencies.props` now, yeah? https://github.com/dotnet/aspnetcore/blob/5782872f7f79628d48dc88f0f60ff249fe589fe6/eng/Dependencies.props#L189-L196

(For my understanding, not asking for a change)

dougbu replied:
No, `%(Version)` metadata doesn't apply to project references and Dependencies.props only contains information about package references.

But also yes, versions for all `@(Reference)` items transformed into `@(PackageReference)` items come from Dependencies.props and have forever. The code in your link is about calculating the `@(LatestPackageReference)` metadata automatically and added support for `%(RTMVersion)` metadata.

wtgodbe replied:
Ah, right, forgot this was a ProjectReference

dougbu committed:
!fixup! Instead remove `$(Razor_NewtonsoftJsonPackageVersion)`

nits:
- remove redundant `%(Reference.Version)` metadata for Microsoft.Css.Parser too
- move `@(Reference)` items together in Microsoft.AspNetCore.Razor.Tools
